DCGAN/train.py

```diff
@@ -0,0 +1,146 @@
+    noise = tf.random.normal([batch_size, noise_dim])
+
+    with tf.GradientTape() as gen_tape, tf.GradientTape() as disc_tape:
+        generated_images = gen_model(noise, training=True)
+        real_output = disc_model(images, training=True)
+        generated_output = disc_model(generated_images, training=True)
+        gen_loss = generator_loss(generated_output)
+        disc_loss = discriminator_loss(real_output, generated_output)
+
+    gen_grads = gen_tape.gradient(gen_loss, gen_model.trainable_variables)
+    disc_grads = disc_tape.gradient(disc_loss, disc_model.trainable_variables)
+
+    generator_optimizer.apply_gradients(zip(gen_grads, gen_model.trainable_variables))
+    discriminator_optimizer.apply_gradients(zip(disc_grads, disc_model.trainable_variables))
+    return gen_loss, disc_loss
+
+def train_model(dataset, epochs, batch_size):
+    step = 0
+    for epoch in range(epochs):
+        for images in dataset:
+            gen_loss, disc_loss = train_step(images, batch_size)
+            if (step+1) % log_frequency == 0:
+                with writer.as_default():
+                    tf.summary.scalar("Generator_Loss", gen_loss, step)
+                    tf.summary.scalar("Discriminator_Loss", disc_loss, step)
+            step = step+1
+            
+        if (epoch + 1) % model_save_frequency == 0:
+            current_time_as_string = f"{datetime.now():%Y-%m-%d_%H:%M:%S}"
+            checkpoint_name = checkpoint_dir + current_time_as_string
+            checkpoint.write(file_prefix = checkpoint_name)
+            print("Checkpoint saved :{}".format(checkpoint_name))
+
+        predictions=gen_model(random_vector, training=False)
+        images=predictions*0.5 + 0.5
+        generate_and_save_images(epoch+1, images)
+        with writer.as_default():
+            tf.summary.image("Generated images", images, step, max_outputs=num_examples)
+
+        print("Epoch {} done".format(epoch+1))
+
+image_save_directory = exp_path +  "/generated_images"
+if not os.path.exists(image_save_directory):
+    mkdir(image_save_directory)
+
+def image_grid(x, size=grid_size):
+    t = tf.unstack(x[:size * size], num=size*size, axis=0)
+    rows = [tf.concat(t[i*size:(i+1)*size], axis=0) 
+            for i in range(size)]
+    image = tf.concat(rows, axis=1)
+    return image
+
+def generate_and_save_images(epoch, images):
+    plt.axis('off')
+    plt.title('Generated Images')
+    if(images.shape[3]==1):
+        plt.imshow(image_grid(images)[:, :, 0], cmap='gray')
+    else:
+        plt.imshow(image_grid(images))
+
+    plt.savefig(image_save_directory + "/sample_image_from_epoch_{:04d}.png".format(epoch))
+    plt.close()
+
+log = logging.getLogger('werkzeug').setLevel(logging.ERROR)
+tb = program.TensorBoard()
+tb.configure(argv=[None, '--logdir', logs_path])
+url = tb.launch()
+print('TensorBoard at {}'.format(url))
+webbrowser.open('http://localhost:6006')
+train_model(dataset, EPOCHS, BATCH_SIZE)
```

Comment:
close tensorboard after training or if someone terminates the program without finishing training

Reply:
TensorBoard automatically closes when the training is finished/the program is closed. However, the chrome tab remains open, and I have no clue on how to close a particular tab from python :P The only workaround is to kill the chrome.exe process but it kills all tabs currently open and therefore is not a good solution